script/copy_deduplicate

```diff
@@ -130,42 +167,87 @@ group.add_argument(
+
 
 def sql_full_table_id(table):
     return table.full_table_id.replace(":", ".")
 
 
-def run_deduplication_query(client, live_table, stable_table, date, dry_run):
+def get_deduplication_query(live_table, hourly):
+    base_template = QUERY_HOUR if hourly else QUERY_DAY
+    base = base_template.format(source_table_spec=sql_full_table_id(live_table))
+    return QUERY_TEMPLATE.format(base=base.strip())
 
-    sql = QUERY_TEMPLATE.format(source_table_spec=sql_full_table_id(live_table))
-    destination = f"{sql_full_table_id(stable_table)}${date:%Y%m%d}"
-    job_config = bigquery.QueryJobConfig(
-        destination=destination,
-        query_parameters=[
-            bigquery.ScalarQueryParameter("submission_date", "DATE", date)
-        ],
-        use_legacy_sql=False,
-        write_disposition=bigquery.WriteDisposition.WRITE_TRUNCATE,
-        dry_run=dry_run,
-    )
-    query_job = client.query(sql, job_config)
-    if dry_run:
-        print(
-            "Would process {} bytes: {}".format(
-                query_job.total_bytes_processed, query_job.to_api_repr()
+
+def get_query_job_configs(client, stable_table, date, dry_run, hourly):
+    kwargs = dict(use_legacy_sql=False, dry_run=dry_run)
+    if hourly:
+        kwargs["allow_large_results"] = True
+        stable_table = client.get_table(stable_table)
+        for hour in range(24):
+            value = (date + timedelta(hours=hour)).isoformat()
+            yield bigquery.QueryJobConfig(
+                destination=get_temporary_table(client),
+                clustering_fields=stable_table.clustering_fields,
+                time_partitioning=stable_table.time_partitioning,
+                query_parameters=[
+                    bigquery.ScalarQueryParameter("submission_hour", "TIMESTAMP", value)
+                ],
+                **kwargs,
             )
-        )
     else:
-        query_job.result()
-        print(
-            "Processed {} bytes to populate {}".format(
-                query_job.total_bytes_processed, destination
-            )
+        yield bigquery.QueryJobConfig(
+            destination=stable_table,
+            write_disposition=bigquery.WriteDisposition.WRITE_TRUNCATE,
+            query_parameters=[
+                bigquery.ScalarQueryParameter("submission_date", "DATE", date)
+            ],
+            **kwargs,
         )
 
 
-def worker_entrypoint(args):
-    run_deduplication_query(*args)
+def run_deduplication_query(client, sql, stable_table, job_config):
+    query_job = client.query(sql, job_config)
+    if not query_job.dry_run:
+        query_job.result()
+    return stable_table, query_job
+
+
+def copy_join_parts(client, stable_table, query_jobs):
+    total_bytes = sum(query.total_bytes_processed for query in query_jobs)
+    if query_jobs[0].dry_run:
+        api_repr = query_jobs[0].to_api_repr()
+        if len(query_jobs) > 1:
+            print(f"Would process {total_bytes} bytes: [{api_repr},...]")
```

Comment:
hourly deduplciation depends on clustering, so total_bytes here is an upper bound. That isn't so bad until you realize that this estimates deduplicating main_summary hourly will scan the full day 24 times and cost $1087.50 😨

Reply:
actual usage for hourly on main_v4 was only 3.3% more than daily, so we should be good.